qbraid_qir/qasm3/visitor.py

```diff
@@ -15,62 +15,305 @@
 # pylint: disable=too-many-instance-attributes,too-many-lines,too-many-branches
 
 """
-Module defining Qasm3 Visitor.
+Module defining Profile-based Qasm3 Visitor for QIR generation.
+
+This refactored approach uses Profile objects to handle different QIR profiles
+without code duplication, based on JSON profile specifications.
 
 """
 import logging
-from typing import Any, Union
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Any, Callable, Dict, List, Optional, Union
 
 import openqasm3.ast as qasm3_ast
 import pyqir
 import pyqir._native
 import pyqir.rt
 from openqasm3.ast import UnaryOperator
+from pyqir import qis
 
 from .elements import QasmQIRModule
 from .exceptions import raise_qasm3_error
-from .maps import map_qasm_op_to_pyqir_callable
+from .maps import PYQIR_ONE_QUBIT_ROTATION_MAP, map_qasm_op_to_pyqir_callable
 
 logger = logging.getLogger(__name__)
 
 
+@dataclass
```

Comment:
It is better to move Lines #43 to #251 to a new `profiles.py` file for better maintainability